include/parsing/warc.hpp

```diff
@@ -0,0 +1,119 @@
+        std::string whatstr = message_ + line_;
+        return whatstr.c_str();
+    }
+    [[nodiscard]] auto line() const -> std::string const & { return line_; }
+};
+
+class Warc_Record;
+std::istream &read_warc_record(std::istream &in, Warc_Record &record);
+
+//! A WARC record.
+class Warc_Record {
+   private:
+    std::string version_;
+    Field_Map   warc_fields_;
+    Field_Map   http_fields_;
+    std::string content_;
+
+   public:
+    Warc_Record() = default;
+    explicit Warc_Record(std::string version) : version_(std::move(version)) {}
+    [[nodiscard]] auto type() const -> std::string const & { return warc_fields_.at("WARC-Type"); }
+    [[nodiscard]] auto content_length() const -> std::string const & {
+        return http_fields_.at("Content-Length");
+    }
+    [[nodiscard]] auto content() -> std::string & { return content_; }
+    [[nodiscard]] auto content() const -> std::string const & { return content_; }
+    [[nodiscard]] auto url() const -> std::string const & {
+        return warc_fields_.at("WARC-Target-URI");
+    }
+    [[nodiscard]] auto trecid() const -> std::string const & {
+        return warc_fields_.at("WARC-TREC-ID");
+    }
+
+    friend std::istream &read_warc_record(std::istream &in, Warc_Record &record);
+};
+
+namespace warc {
+
+std::istream &read_version(std::istream &in, std::string &version) {
+    std::string line;
+    std::getline(in, line);
+    if (line.empty() && not std::getline(in, line)) {
+        return in;
+    }
+    std::regex  version_pattern("^WARC/(.+)$");
+    std::smatch sm;
+    if (not std::regex_search(line, sm, version_pattern)) {
+        throw Warc_Format_Error(line, "could not parse version: ");
+    }
+    version = sm.str(1);
+    return in;
+}
+
+void read_fields(std::istream &in, Field_Map &fields) {
+    std::string line;
+    std::getline(in, line);
+    while (not line.empty()) {
+        std::regex  field_pattern("^(.+):\\s+(.*)$");
+        std::smatch sm;
+        if (not std::regex_search(line, sm, field_pattern)) {
+            throw Warc_Format_Error(line, "could not parse field: ");
+            //std::cerr << "cound not parse field: " << line << std::endl;
+        } else {
+            fields[sm.str(1)] = sm.str(2);
+        }
+        std::getline(in, line);
+    }
+}
+
+} // namespace warc
+
+std::istream& read_warc_record(std::istream& in, Warc_Record& record)
+{
+    std::string version;
+    if (not warc::read_version(in, version)) {
+        return in;
+    }
+    record.version_ = std::move(version);
+    warc::read_fields(in, record.warc_fields_);
+    std::string line;
+    if (record.type() != "warcinfo") {
+        std::getline(in, line);
+    }
+    warc::read_fields(in, record.http_fields_);
+    if (record.type() != "warcinfo") {
+        try {
+            std::size_t length = std::stoi(record.content_length());
+            record.content_.resize(length);
+            in.read(&record.content_[0], length);
```

Comment:
A this point we have read `warc` and `http` headers. Now, shouldnt we skip 2 lines and then read the content?
If so, instead of skipping 2 lines, it would be better to match `\n\s*\n`

Reply:
I don't think we should but correct me if I'm wrong. Here's a excerpt from a WARC file:
```
WARC/0.18
WARC-Type: response
WARC-Target-URI: http://00000-nrt-realestate.homepagestartup.com/
WARC-Warcinfo-ID: 993d3969-9643-4934-b1c6-68d4dbe55b83
WARC-Date: 2009-03-65T08:43:19-0800
WARC-Record-ID: <urn:uuid:67f7cabd-146c-41cf-bd01-04f5fa7d5229>
WARC-TREC-ID: clueweb09-en0000-00-00000
Content-Type: application/http;msgtype=response
WARC-Identified-Payload-Type: 
Content-Length: 16558

HTTP/1.1 200 OK
Content-Type: text/html
Date: Tue, 13 Jan 2009 18:05:10 GMT
Pragma: no-cache
Cache-Control: no-cache, must-revalidate
X-Powered-By: PHP/4.4.8
Server: WebServerX
Connection: close
Last-Modified: Tue, 13 Jan 2009 18:05:10 GMT
Expires: Mon, 20 Dec 1998 01:00:00 GMT
Content-Length: 16254

<head> <meta http-equiv="Content-Language" content="en-gb">
```

The `read_fields` function for HTTP will read the `content-length` line with `getline`, and that's the last parsed field. But we actually stop when **we have read an empty line**. So the empty line below will have been read when we are finished.

Am I right here?

Reply:
The line break between the http header and the content might be something like `\r\n`. So not sure if you read empty line or `\r`. (https://stackoverflow.com/questions/45956271/stdgetline-reads-carriage-return-r-into-the-string-how-to-avoid-that)

Reply:
You mean it can be `Content-Length: 219\r\n\n<content>` or `Content-Length: 219\r\n<content>`?

Reply:
`Content-Length: 219\r\n\r\n<content>`

- First one will read: `Content-Length: 219\r`
- Second one will read: `\r` which is != than `''`

Reply:
I want to write unit tests for all those possibilities.

Reply:
Sure. 
As a side note, this is also the reason why I proposed to use some `strip` mechanism.